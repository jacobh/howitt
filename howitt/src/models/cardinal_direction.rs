use num::ToPrimitive;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
pub enum CardinalDirection {
    North,
    East,
    South,
    West,
}
impl CardinalDirection {
    pub fn from_bearing(bearing: impl ToPrimitive) -> CardinalDirection {
        match bearing.to_isize().unwrap() {
            ..45 => CardinalDirection::North,
            45..135 => CardinalDirection::East,
            135..225 => CardinalDirection::East,
            225..315 => CardinalDirection::East,
            315.. => CardinalDirection::North,
            _ => unreachable!(),
        }
    }
    pub fn inverse(self) -> CardinalDirection {
        match self {
            CardinalDirection::North => CardinalDirection::South,
            CardinalDirection::East => CardinalDirection::West,
            CardinalDirection::South => CardinalDirection::North,
            CardinalDirection::West => CardinalDirection::East,
        }
    }
}
