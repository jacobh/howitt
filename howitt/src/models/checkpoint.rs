use std::{fmt::Display, str::FromStr};

use anyhow::anyhow;
use chrono::DateTime;
use serde::{Deserialize, Serialize};
use thiserror::Error;

use crate::{ext::ulid::generate_ulid, models::Model};

use super::IndexModel;

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum CheckpointType {
    RailwayStation,
    Hut,
    Locality,
    Generic,
}

impl FromStr for CheckpointType {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "RAILWAY_STATION" => Ok(CheckpointType::RailwayStation),
            "HUT" => Ok(CheckpointType::Hut),
            "LOCALITY" => Ok(CheckpointType::Locality),
            "GENERIC" => Ok(CheckpointType::Generic),
            _ => Err(()),
        }
    }
}

impl Display for CheckpointType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s = match self {
            CheckpointType::RailwayStation => "RAILWAY_STATION",
            CheckpointType::Hut => "HUT",
            CheckpointType::Locality => "LOCALITY",
            CheckpointType::Generic => "GENERIC",
        };

        f.write_str(s)
    }
}

crate::model_id!(CheckpointId, "CHECKPOINT");

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct Checkpoint {
    pub id: ulid::Ulid,
    pub name: String,
    #[serde(with = "crate::ext::serde::point_tuple")]
    pub point: geo::Point<f64>,
    pub checkpoint_type: CheckpointType,
}

impl IndexModel for Checkpoint {
    type Id = CheckpointId;

    fn id(&self) -> CheckpointId {
        CheckpointId::from(self.id)
    }
}

#[derive(Debug, Error)]
#[error("Checkpoint conversion failed")]
pub enum CheckpointError {
    MissingName,
}

impl TryFrom<gpx::Waypoint> for Checkpoint {
    type Error = CheckpointError;
    fn try_from(value: gpx::Waypoint) -> Result<Self, Self::Error> {
        let waypoint_created_at = value
            .time
            .map(|time| time.format())
            .transpose()
            .unwrap()
            .as_deref()
            .map(DateTime::parse_from_rfc3339)
            .transpose()
            .unwrap();

        let id = generate_ulid(waypoint_created_at, &value).unwrap();

        match value.name.clone() {
            Some(name) => Ok(Checkpoint {
                id,
                name,
                point: value.point(),
                checkpoint_type: value
                    ._type
                    .unwrap_or("".to_string())
                    .parse()
                    .unwrap_or(CheckpointType::Generic),
            }),
            None => Err(CheckpointError::MissingName),
        }
    }
}
