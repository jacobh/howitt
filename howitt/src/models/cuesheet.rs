use serde::{Deserialize, Serialize};

use super::point_of_interest::PointOfInterest;

#[derive(Debug, Serialize, Deserialize)]
pub struct Cuesheet {
    pub cues: Vec<Cue>,
}

#[derive(Debug, Serialize, Deserialize)]
pub enum CueStop {
    Start,
    End,
    POI(PointOfInterest),
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Cue {
    pub origin: CueStop,
    pub destination: CueStop,
    pub distance_m: f64,
    pub vertical_ascent_m: Option<f64>,
    pub vertical_descent_m: Option<f64>,
}
