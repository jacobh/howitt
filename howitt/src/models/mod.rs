use std::{
    hash::Hash,
    marker::{ConstParamTy, PhantomData},
    sync::Arc,
};

use futures::FutureExt;
use serde::{Deserialize, Serialize};
use tokio::sync::OnceCell;
use uuid::Timestamp;

use crate::repos::AnyhowRepo;

pub mod cardinal_direction;
pub mod config;
pub mod cuesheet;
pub mod external_ref;
pub mod filters;
pub mod maybe_pair;
pub mod point;
pub mod point_of_interest;
pub mod ride;
pub mod route;
pub mod route_description;
pub mod segment;
pub mod segment_summary;
pub mod slope_end;
pub mod tag;
pub mod terminus;
pub mod trip;
pub mod user;

pub trait Model: Send + Sync + Sized + Clone + 'static {
    type Id: ModelId;
    type IndexItem: IndexItem<Id = Self::Id>;
    type Filter: Send + Sync + Sized + Clone + 'static;

    fn model_name() -> &'static str {
        Self::Id::model_name()
    }
    fn id(&self) -> Self::Id;

    fn as_index(&self) -> &Self::IndexItem;
}

pub trait IndexModel {
    type Id: ModelId;
    type Filter: Send + Sync + Sized + Clone + 'static;

    fn id(&self) -> Self::Id;
}

impl<T, ID> IndexItem for T
where
    T: IndexModel<Id = ID> + 'static + Send + Sync + Sized + Clone,
    ID: ModelId,
{
    type Id = ID;

    fn model_id(&self) -> Self::Id {
        self.id()
    }
}

impl<T, ID, F> Model for T
where
    T: IndexModel<Id = ID, Filter = F> + Send + Sync + Sized + Clone + 'static,
    ID: ModelId + 'static,
    F: Send + Sync + Sized + Clone + 'static,
{
    type Id = ID;

    type IndexItem = T;

    type Filter = F;

    fn id(&self) -> Self::Id {
        self.id()
    }

    fn as_index(&self) -> &Self::IndexItem {
        self
    }
}

pub trait IndexItem: 'static + Send + Sync + Clone {
    type Id: ModelId;

    fn model_id(&self) -> Self::Id;
}

pub trait OtherItem: 'static + Send + Sync + Clone {
    type Id: ModelId;

    fn model_id(&self) -> Self::Id;
    fn item_name(&self) -> String;
    fn item_id(&self) -> String;
}

#[derive(Clone)]
enum ModelRefInitial<M: Model> {
    Model(M),
    Index(M::IndexItem),
}

pub struct ModelRefInner<M: Model> {
    initial: ModelRefInitial<M>,
    model: OnceCell<Result<Arc<M>, Arc<anyhow::Error>>>,
}

#[derive(Clone)]
pub struct ModelRef<M: Model> {
    inner: Arc<ModelRefInner<M>>,
}

impl<M> ModelRef<M>
where
    M: Model + Clone,
{
    fn new(initial: ModelRefInitial<M>) -> ModelRef<M> {
        ModelRef {
            inner: Arc::new(ModelRefInner {
                initial,
                model: OnceCell::new(),
            }),
        }
    }
    pub fn from_model(model: M) -> ModelRef<M> {
        ModelRef::new(ModelRefInitial::Model(model))
    }
    pub fn from_index(index: M::IndexItem) -> ModelRef<M> {
        ModelRef::new(ModelRefInitial::Index(index))
    }
    pub fn id(&self) -> M::Id {
        match &self.inner.initial {
            ModelRefInitial::Model(model) => model.id(),
            ModelRefInitial::Index(index) => index.model_id(),
        }
    }
    pub fn as_index(&self) -> &M::IndexItem {
        match &self.inner.initial {
            ModelRefInitial::Model(model) => model.as_index(),
            ModelRefInitial::Index(index) => index,
        }
    }
    pub async fn to_model<R: AsRef<dyn AnyhowRepo<Model = M>> + Send + Sync>(
        &self,
        repo: R,
    ) -> Result<Arc<M>, Arc<anyhow::Error>> {
        self.inner
            .model
            .get_or_init(move || match self.inner.initial.clone() {
                ModelRefInitial::Model(model) => {
                    futures::future::ready(Ok(Arc::new(model))).boxed()
                }
                ModelRefInitial::Index(index) => (async move {
                    repo.as_ref()
                        .get(index.model_id())
                        .await
                        .map(Arc::new)
                        .map_err(Arc::new)
                })
                .boxed(),
            })
            .await
            .as_ref()
            .map(Clone::clone)
            .map_err(Clone::clone)
    }
}

#[derive(Clone)]
pub struct EmptyOtherItem<ID>(PhantomData<ID>);

impl<ID> OtherItem for EmptyOtherItem<ID>
where
    ID: ModelId + 'static,
{
    type Id = ID;

    fn model_id(&self) -> Self::Id {
        todo!()
    }

    fn item_name(&self) -> String {
        todo!()
    }

    fn item_id(&self) -> String {
        todo!()
    }
}

pub trait ModelId:
    Send + Sync + std::fmt::Debug + std::fmt::Display + PartialEq + Copy + Clone + Hash + Eq + 'static
{
    fn model_name() -> &'static str;
}

#[derive(ConstParamTy, PartialEq, Eq, Clone, Copy)]
pub enum ModelName {
    Photo,
    Checkpoint,
    Ride,
    Route,
    Segment,
    User,
    Trip,
}
impl ModelName {
    const fn to_str(self) -> &'static str {
        match self {
            ModelName::Photo => "PHOTO",
            ModelName::Checkpoint => "CHECKPOINT",
            ModelName::Ride => "RIDE",
            ModelName::Route => "ROUTE",
            ModelName::Segment => "SEGMENT",
            ModelName::User => "USER",
            ModelName::Trip => "TRIP",
        }
    }
}

#[derive(derive_more::From, derive_more::Into, PartialEq, Eq, Hash, Clone, Copy)]
pub struct ModelUuid<const NAME: ModelName>(uuid::Uuid);

impl<const NAME: ModelName> ModelUuid<NAME> {
    pub fn new() -> ModelUuid<NAME> {
        ModelUuid::<NAME>(uuid::Uuid::now_v7())
    }
    pub fn as_uuid(&self) -> &uuid::Uuid {
        &self.0
    }
    pub fn from_datetime(datetime: chrono::DateTime<chrono::Utc>) -> ModelUuid<NAME> {
        ModelUuid::<NAME>(uuid::Uuid::new_v7(
            Timestamp::from_unix(
                uuid::timestamp::context::ContextV7::new(),
                datetime.timestamp() as u64,
                0,
            )
            .into(),
        ))
    }
    pub fn get_or_from_datetime(
        id: Option<ModelUuid<NAME>>,
        datetime: &chrono::DateTime<chrono::Utc>,
    ) -> ModelUuid<NAME> {
        match id {
            Some(id) => id,
            None => ModelUuid::<NAME>::from_datetime(*datetime),
        }
    }
}

impl<const NAME: ModelName> ModelId for ModelUuid<NAME> {
    fn model_name() -> &'static str {
        NAME.to_str()
    }
}

impl<const NAME: ModelName> std::fmt::Debug for ModelUuid<NAME> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}#{}", NAME.to_str(), self.0)
    }
}

impl<const NAME: ModelName> std::fmt::Display for ModelUuid<NAME> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}#{}", NAME.to_str(), self.0)
    }
}

impl<const NAME: ModelName> Default for ModelUuid<NAME> {
    fn default() -> Self {
        Self::new()
    }
}

impl<const NAME: ModelName> Serialize for ModelUuid<NAME> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        self.to_string().serialize(serializer)
    }
}

impl<'de, const NAME: ModelName> Deserialize<'de> for ModelUuid<NAME> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let model_id: String = Deserialize::deserialize(deserializer)?;
        let parts = model_id.split('#').collect::<Vec<_>>();

        if parts.len() != 2 {
            return Err(serde::de::Error::custom("expected 2 parts"));
        }

        let name = parts[0];
        let id = parts[1];

        if name != NAME.to_str() {
            return Err(serde::de::Error::custom(
                "model name component of ID did not match",
            ));
        }

        std::str::FromStr::from_str(id)
            .map(ModelUuid::<NAME>)
            .map_err(serde::de::Error::custom)
    }
}

// #[cfg(test)]
// mod tests {
//     use chrono::{DateTime, Utc};
//     use test_case::test_case;

//     use crate::models::route::RouteId;

//     fn datetime1() -> DateTime<Utc> {
//         DateTime::parse_from_rfc3339("2023-01-01T00:00:00Z")
//             .unwrap()
//             .into()
//     }

//     fn datetime2() -> DateTime<Utc> {
//         DateTime::parse_from_rfc3339("2023-10-01T00:00:00Z")
//             .unwrap()
//             .into()
//     }

//     const ULID_PREFIX1: &str = "ROUTE#01GNNA1J00";
//     const ULID_PREFIX2: &str = "ROUTE#01HBM8HS00";

//     #[test_case(Some(RouteId::from_datetime(datetime1())), datetime2(), ULID_PREFIX1)]
//     #[test_case(Some(RouteId::from_datetime(datetime2())), datetime1(), ULID_PREFIX2)]
//     #[test_case(None, datetime2(), ULID_PREFIX2)]
//     fn test_get_or_from_datetime(
//         existing_id: Option<RouteId>,
//         datetime: DateTime<Utc>,
//         expected_prefix: &str,
//     ) {
//         let id = RouteId::get_or_from_datetime(existing_id, &datetime).to_string();
//         let (prefix, _) = id.split_at(expected_prefix.len());

//         assert_eq!(expected_prefix, prefix);
//     }
// }
