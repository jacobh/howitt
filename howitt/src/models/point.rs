use chrono::{DateTime, TimeZone, Utc};
use geo::{CoordsIter, GeodesicBearing, LineString, Simplify};
use itertools::Itertools;
use serde::{Deserialize, Serialize};

use super::ModelId;

pub trait Point: std::fmt::Debug + Clone {
    fn as_geo_point(&self) -> &geo::Point;
    fn elevation_meters(&self) -> Option<&f64>;
    fn to_elevation_point(&self) -> Option<ElevationPoint>;

    fn x_y(&self) -> (f64, f64) {
        geo::Point::x_y(*self.as_geo_point())
    }

    fn ordered_x_y(
        &self,
    ) -> (
        ordered_float::OrderedFloat<f64>,
        ordered_float::OrderedFloat<f64>,
    ) {
        let (x, y) = self.x_y();

        (
            ordered_float::OrderedFloat(x),
            ordered_float::OrderedFloat(y),
        )
    }

    fn x_y_z(&self) -> (f64, f64, Option<f64>) {
        let (x, y) = self.x_y();

        (x, y, self.elevation_meters().copied())
    }

    fn to_x_y_vec(&self) -> Vec<f64> {
        let (x, y) = self.x_y();
        vec![x, y]
    }

    fn into_x_y_vec(self) -> Vec<f64> {
        let (x, y) = self.x_y();
        vec![x, y]
    }
}

impl Point for geo::Point {
    fn as_geo_point(&self) -> &geo::Point {
        self
    }

    fn elevation_meters(&self) -> Option<&f64> {
        None
    }

    fn to_elevation_point(&self) -> Option<ElevationPoint> {
        None
    }
}

impl<'a> Point for &'a geo::Point {
    fn as_geo_point(&self) -> &geo::Point {
        self
    }

    fn elevation_meters(&self) -> Option<&f64> {
        None
    }

    fn to_elevation_point(&self) -> Option<ElevationPoint> {
        None
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct ElevationPoint {
    pub point: geo::Point,
    pub elevation: f64,
}

impl Serialize for ElevationPoint {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        (self.point.x(), self.point.y(), self.elevation).serialize(serializer)
    }
}

impl<'de> Deserialize<'de> for ElevationPoint {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let (x, y, elevation) = Deserialize::deserialize(deserializer)?;
        Ok(ElevationPoint {
            point: geo::Point::new(x, y),
            elevation,
        })
    }
}

impl Point for ElevationPoint {
    fn as_geo_point(&self) -> &geo::Point {
        &self.point
    }

    fn elevation_meters(&self) -> Option<&f64> {
        Some(&self.elevation)
    }

    fn to_elevation_point(&self) -> Option<ElevationPoint> {
        Some(self.clone())
    }
}

impl<'a> Point for &'a ElevationPoint {
    fn as_geo_point(&self) -> &geo::Point {
        &self.point
    }

    fn elevation_meters(&self) -> Option<&f64> {
        Some(&self.elevation)
    }

    fn to_elevation_point(&self) -> Option<ElevationPoint> {
        Some((*self).clone())
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct TemporalElevationPoint {
    pub datetime: DateTime<Utc>,
    pub point: geo::Point,
    pub elevation: f64,
}

impl Serialize for TemporalElevationPoint {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        (
            self.datetime.timestamp(),
            self.point.x(),
            self.point.y(),
            self.elevation,
        )
            .serialize(serializer)
    }
}

impl<'de> Deserialize<'de> for TemporalElevationPoint {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let (timestamp, x, y, elevation) = Deserialize::deserialize(deserializer)?;
        Ok(TemporalElevationPoint {
            datetime: Utc
                .timestamp_opt(timestamp, 0)
                .single()
                .ok_or_else(|| serde::de::Error::custom("couldnt parse timestamp"))?,
            point: geo::Point::new(x, y),
            elevation,
        })
    }
}

impl Point for TemporalElevationPoint {
    fn as_geo_point(&self) -> &geo::Point {
        &self.point
    }

    fn elevation_meters(&self) -> Option<&f64> {
        Some(&self.elevation)
    }

    fn to_elevation_point(&self) -> Option<ElevationPoint> {
        Some(ElevationPoint {
            point: self.point,
            elevation: self.elevation,
        })
    }
}

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct PointChunk<ID, P> {
    pub model_id: ID,
    pub idx: usize,
    pub points: Vec<P>,
}
impl<ID, P> PointChunk<ID, P>
where
    ID: ModelId,
{
    pub fn new_chunks(model_id: ID, points: impl IntoIterator<Item = P>) -> Vec<PointChunk<ID, P>> {
        points
            .into_iter()
            .chunks(2500)
            .into_iter()
            .enumerate()
            .map(|(idx, points)| PointChunk {
                model_id,
                idx,
                points: points.collect(),
            })
            .collect()
    }
    pub fn iter_points(chunks: &[PointChunk<ID, P>]) -> impl Iterator<Item = &P> + '_ {
        chunks.iter().flat_map(|chunk| chunk.points.iter())
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct PointDelta {
    pub distance: f64,
    pub bearing: f64,
    pub elevation_gain: Option<f64>,
}

impl PointDelta {
    pub fn zero() -> PointDelta {
        PointDelta {
            distance: 0.0,
            bearing: 0.0,
            elevation_gain: None,
        }
    }

    pub fn from_points<P1: Point, P2: Point>(p1: &P1, p2: &P2) -> PointDelta {
        let (bearing, distance) = p1
            .as_geo_point()
            .geodesic_bearing_distance(*p2.as_geo_point());

        let bearing = bearing % 360.0;

        let elevation_gain = match (p1.elevation_meters(), p2.elevation_meters()) {
            (Some(e1), Some(e2)) => Some(e2 - e1),
            _ => None,
        };

        PointDelta {
            distance,
            bearing,
            elevation_gain,
        }
    }
    pub fn from_points_tuple<P: Point>((p1, p2): (&P, &P)) -> PointDelta {
        PointDelta::from_points(p1, p2)
    }
}

pub fn generate_point_deltas<'a, P: Point + 'a>(
    points: impl IntoIterator<Item = &'a P>,
) -> Vec<PointDelta> {
    let mut is_points_empty = true;

    let deltas = points
        .into_iter()
        .inspect(|_| is_points_empty = false)
        .tuple_windows()
        .map(PointDelta::from_points_tuple)
        .collect_vec();

    // add an empty delta to the start to keep this aligned with the input
    if !is_points_empty {
        [vec![PointDelta::zero()], deltas].concat()
    } else {
        deltas
    }
}

// const LOWER_EPSILON: f64 = 0.0;
// const UPPER_EPSILON: f64 = 0.001;

// #[derive(Debug, Clone, Copy)]
// pub enum UpperEpsilon {
//     Attempts(usize),
//     Value(f64),
// }
// impl UpperEpsilon {
//     fn value(&self) -> f64 {
//         match self {
//             UpperEpsilon::Attempts(i) => ,
//             UpperEpsilon::Value(x) => *x,
//         }
//     }
//     fn increment_attempts(self) -> UpperEpsilon {
//         match self {
//             UpperEpsilon::Attempts(i) => UpperEpsilon::Attempts(i + 1),
//             UpperEpsilon::Value(y) => UpperEpsilon::Value(y),
//         }
//     }
// }

pub struct SimplifyState {
    epsilon: (f64, Option<f64>),
    i: usize,
    oversimplified: Option<LineString>,
}

impl Default for SimplifyState {
    fn default() -> Self {
        Self {
            epsilon: (0.0, None),
            oversimplified: Default::default(),
            i: 0,
        }
    }
}

pub fn simplify_linestring(
    linestring: LineString,
    max_points: usize,
    state: Option<SimplifyState>,
) -> LineString {
    if linestring.coords_count() <= max_points {
        return linestring;
    }

    let SimplifyState {
        epsilon: (lower, upper),
        i,
        oversimplified,
    } = state.unwrap_or_default();

    let epsilon = (lower + upper.unwrap_or(0.0005 * f64::powi(2.0, (i + 1) as i32))) / 2.0;

    let simplified = Simplify::simplify(&linestring, &epsilon);
    let count = simplified.coords_count();

    if count == max_points {
        // bang on target, return simplified
        simplified
    } else if i >= 50 {
        oversimplified.unwrap_or(simplified)
    } else if count > max_points {
        // too many points
        simplify_linestring(
            linestring,
            max_points,
            Some(SimplifyState {
                epsilon: (epsilon, upper),
                oversimplified,
                i: i + 1,
            }),
        )
    } else {
        if i > 0 {
            let oversimplified = oversimplified.unwrap_or(LineString::new(vec![]));
            let oversimplified = Some(
                if simplified.coords_count() > oversimplified.coords_count() {
                    simplified
                } else {
                    oversimplified
                },
            );

            // not enough points and we're not using the default epsilon. search our way back to the limit
            simplify_linestring(
                linestring,
                max_points,
                Some(SimplifyState {
                    epsilon: (lower, Some(epsilon)),
                    i: i + 1,
                    oversimplified,
                }),
            )
        } else {
            // initial epsilon was good enough to get us below the limit
            simplified
        }
    }
}

pub fn simplify_points<P: Point>(points: &[P], target_points: usize) -> Vec<P> {
    let linestring = LineString::from_iter(points.iter().map(|p| *p.as_geo_point()));

    let simplified = simplify_linestring(linestring, target_points, None);

    simplified
        .points()
        .map(|point| points.iter().find(|p| p.as_geo_point() == &point))
        .flatten()
        .cloned()
        .collect()
}
