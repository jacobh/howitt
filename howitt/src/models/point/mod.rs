pub mod closest_point;
pub mod delta2;
pub mod elevation_point;
pub mod point;
pub mod point_chunk;
pub mod progress;
pub mod temporal_elevation_point;

pub use closest_point::*;
pub use elevation_point::*;
pub use point::*;
pub use point_chunk::*;
pub use temporal_elevation_point::*;
