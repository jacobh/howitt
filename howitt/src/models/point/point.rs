use super::point_delta::DeltaData;

pub trait Point: std::fmt::Debug + Clone {
    type DeltaData: DeltaData;

    fn as_geo_point(&self) -> &geo::Point;
    fn delta(&self, other: &Self) -> Self::DeltaData;

    fn x_y(&self) -> (f64, f64) {
        geo::Point::x_y(*self.as_geo_point())
    }

    fn ordered_x_y(
        &self,
    ) -> (
        ordered_float::OrderedFloat<f64>,
        ordered_float::OrderedFloat<f64>,
    ) {
        let (x, y) = self.x_y();

        (
            ordered_float::OrderedFloat(x),
            ordered_float::OrderedFloat(y),
        )
    }

    fn to_x_y_vec(&self) -> Vec<f64> {
        let (x, y) = self.x_y();
        vec![x, y]
    }

    fn into_x_y_vec(self) -> Vec<f64> {
        let (x, y) = self.x_y();
        vec![x, y]
    }
}

impl Point for geo::Point {
    type DeltaData = ();

    fn as_geo_point(&self) -> &geo::Point {
        self
    }

    fn delta(&self, _: &Self) -> Self::DeltaData {}
}

impl Point for &geo::Point {
    type DeltaData = ();

    fn as_geo_point(&self) -> &geo::Point {
        self
    }

    fn delta(&self, _: &Self) -> Self::DeltaData {}
}
