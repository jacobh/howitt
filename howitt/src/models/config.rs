use serde::{Deserialize, Serialize};

use crate::models::{route::RouteId, IndexModel, ModelId};

#[derive(PartialEq, Copy, Clone)]
pub struct ConfigId;

impl std::fmt::Display for ConfigId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "CONFIG#SINGLETON")
    }
}

impl ModelId for ConfigId {
    fn model_name() -> &'static str {
        "CONFIG"
    }
}

#[derive(Serialize, Deserialize, Default, Clone)]
pub struct Config {
    pub starred_route_ids: Vec<RouteId>,
}

impl IndexModel for Config {
    type Id = ConfigId;

    fn id(&self) -> ConfigId {
        ConfigId
    }
}
