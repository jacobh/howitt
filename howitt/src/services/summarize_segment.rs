use itertools::Itertools;

use crate::models::{
    point::{Point, PointDelta},
    segment_summary::SegmentSummary,
};

use thiserror::Error;

#[derive(Error, Debug, Clone)]
#[error("Failed to summarize segment")]
pub enum SummarizeError {
    NotEnoughPoints,
}

pub fn summarize_segment<P: Point>(points: &[P]) -> Result<SegmentSummary, SummarizeError> {
    if points.len() < 2 {
        return Err(SummarizeError::NotEnoughPoints);
    }

    let summary = SegmentSummary {
        distance_m: 0.0,
        elevation: None,
    };

    let summary = points
        .iter()
        .tuple_windows()
        .map(PointDelta::from_points_tuple)
        // .flatten()
        .fold::<SegmentSummary, _>(
            summary,
            |mut summary,
             PointDelta {
                 distance,
                 elevation_gain,
                 ..
             }| {
                summary.distance_m += distance;

                if let Some(elevation_gain) = elevation_gain {
                    let mut elevation_summary = summary.elevation.unwrap_or_default();

                    if elevation_gain > 0.0 {
                        elevation_summary.elevation_ascent_m += elevation_gain;
                    } else {
                        elevation_summary.elevation_descent_m += elevation_gain.abs();
                    }

                    summary.elevation = Some(elevation_summary);
                }

                summary
            },
        );

    Ok(SegmentSummary {
        distance_m: f64::round(summary.distance_m * 100.0) / 100.0,
        ..summary
    })
}

#[cfg(test)]
mod tests {
    use crate::models::point::ElevationPoint;

    use super::*;

    #[test]
    fn it_succeeds_for_two_points() {
        let point1 = ElevationPoint {
            point: geo::Point::new(146.60587, -37.2154),
            elevation: 1100.0,
        };
        let point2 = ElevationPoint {
            point: geo::Point::new(146.68021, -37.20515),
            elevation: 1400.0,
        };

        let result =
            summarize_segment(&[point1.clone(), point2.clone()]).expect("summarize worked");

        insta::assert_debug_snapshot!(result);
    }
}
