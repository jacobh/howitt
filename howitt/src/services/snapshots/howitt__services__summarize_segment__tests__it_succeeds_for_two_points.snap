---
source: howitt/src/services/summarize_segment.rs
expression: result
---
SegmentSummary {
    distance_m: 6696.19,
    elevation: Some(
        ElevationSummary {
            elevation_ascent_m: 300.0,
            elevation_descent_m: 0.0,
        },
    ),
}
